### ISSUE TITLE
100% GPU load on empty scene

### ISSUE THREAD

**Issue by FyiurAmron** [godotengine/godot#60967]
### Godot version

3.x (multiple)

### System information

Windows 7, Radeon RX 580, Adrenalin 21.5.2, GLES2/3

### Issue description

Empty scene in Godot 3.x (it's happening for me since at least 3.2, and is still happening with 3.5) causes 100% GPU load. Adding a sensible `run/frame_delay_msec` prevents immediate GPU inferno (GPU still bursts to 100% every frame though), but also (obviously) induces stutter etc.

At first I thought that this might be a VSync issue, but no VSync setting (neither via Godot project settings nor GPU settings) helped. Manually checking the amount of rendered frames per sec suggests that VSync is indeed on.

It doesn't happen for Godot 4.x at all, FWIW. I would just move to Godot 4, but my project is C#-based ;D

### Steps to reproduce

create empty scene, run, watch how the GPU spews fire

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/60100.

**Comment by FyiurAmron** [comment]
@Calinou I'm quite convinced this is related to some specific driver/OS config (maybe Windows 7 + Radeon?) going haywire on otherwise sound code/shaders etc. Trying to run Godot 3.x in a Linux VM (Ubuntu @ VirtualBox) results in `Built-in function "texelFetch(sampler2D, ivec2, int)" is only supported on the GLES3 backend, but your project is using GLES2.` - and trying to switch to GLES3 does nothing (editor reverts to GLES2 again).

**Force FPS** set to current VSync Hz didn't do anything at all, but **Low Processor Mode** with default 6900 usec setting **did** reduce the GPU load to 0%, although the GPU clocks were still set to high-performance mode. On 4.x it still gets to almost idle (near the lowest possible) clocking without any tricks.

**Comment by Calinou** [comment]
> BTW, I was wondering if there's some way to profile/debug the GPU-intensive parts of Godot without recompiling it from source (doing that fails on my box, due to some Python-not-supported-on-Win7-related etc. stuff).

You can use a GPU profiler, but these are specific to your brand of GPU and generally only support recent hardware (especially for Vulkan profiling). OpenGL profiling is also often only available in limited form, since GPU profilers are typically focused around DirectX support.

**Comment by FyiurAmron** [comment]
I've noticed there are newer GPU drivers for my platform, but updating from 21.5.2 to 22.6.1 didn't fix this. I'll try repro-ing on another Win box I got my hands on lately (Win10 laptop) and see how it goes there.

**Comment by FyiurAmron** [comment]
@Calinou I've found one thing, hard to say for me if it's expected: if I press the mouse on the "close window" X button the GPU usage drops to 0% for as long as I keep the mouse button pressed. It doesn't matter if I keep the mouse cursor over the X button, move it over the game window, move it completely outside etc., as long as I keep the button pressed. Moving the window, resizing it etc. doesn't seem to trigger this however.

**Comment by rintroix** [comment]
Can confirm 100% CPU usage on one core on empty project on 3.5 stable Godot from flatpak.

    fedora linux 36, both Wayland and X11 Gnome sessions.
    Godot Engine v3.5.stable.flathub.991bb6ac7
    OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 2000 (SNB GT1)

Bug is not present with manually downloaded Godot 3.5 from the web site.

**Comment by Calinou** [comment]
> Can confirm 100% CPU usage on one core on empty project on 3.5 stable Godot from flatpak.

This is likely a different issue (caused by something Flatpak is doing or one of its libraries), and is tracked at https://github.com/flathub/org.godotengine.Godot/issues/108. I don't know why it's happening.

**Comment by FyiurAmron** [comment]
@Calinou correct me if I'm wrong, but wasn't 22.6.1 the last release for Win7? 22.7.1 & 22.8.1 seem to be only in Win10/11 versions, as far as I can see.

**Comment by Calinou** [comment]
> @Calinou correct me if I'm wrong, but wasn't 22.6.1 the last release for Win7? 22.7.1 & 22.8.1 seem to be only in Win10/11 versions, as far as I can see.

Windows 7 is EOL, so GPU manufacturers are no longer releasing new drivers for it.

**Comment by FyiurAmron** [comment]
@Calinou out of curiosity: I wasn't able to find any place that would explicitly say that Godot does or doesn't support Win7 (e.g. Unity explicitly says it _still_ supports Win7, UE4 also say it does in 4.27, UE5 says it doesn't and that it requires full DX11 explicitly) - what is actually the case here ATM?

**Comment by Calinou** [comment]
>  - what is actually the case here ATM?

Godot still explicitly support Windows 7. We aim to have Godot 4.0 support Windows 7 too, but third-party changes may prevent this from materializing. It's not guaranteed that releases past 4.0.x will keep Windows 7 support.

**Comment by SirTodd** [comment]
I developed a Control node only game on Godot and I get many complaints on mobile for high battery usage. On pc the GPU usage starts at 4% and goes up to 20%, sometimes even 40%, with a complete empty scene, capped at 60 fps, both OpenGL 2 and 3.

I plan to make a nother Control node game, but I simply cannot develop using Godot while this problem remains.

I have a RTX 3070 TI, and I would get the same results on a GTX 960 btw.

**Comment by XDRiVE888** [comment]
I CONFIRM this bug, with the only difference being that on my Windows 10 device, Intel HD Graphics 2000, the GPU loads a 50% empty Godot 3.5.1 project. I noticed that the load is proportional to the size of the window, that is, the smaller the window, the less load on the gpu. This is the same in html5 project too, although the load is a little more - 65%.
I consider this to be a significant performance bug, as it limits the game's capabilities, and on a mobile device causes a rapid discharge. It seems that somewhere there is an unoptimized copying of data to the screen or gpu. For comparison, on my full hd device, a 60 fps video displayed in the browser loads the gpu by 35% of everything. It turns out that the empty Godot project loads the gpu almost twice as much as playing a 60 fps video in full hd. I received data on the load on the gpu using the GPU-Z utility.
And yet, perhaps this will help with the search for this bug, then when you set the get_viewport().render_target_update_mode = 0 code, the load on the GPU completely disappears to 0%, well, it’s clear that nothing will be drawn and displayed with this code.

**Comment by Calinou** [comment]
> Intel HD Graphics 2000, the GPU loads a 50% empty Godot 3.5.1 project

This is a very slow IGP – it's expected to show high load in basic 3D scenes.

**Comment by XDRiVE888** [comment]
> > Targeting 60 FPS on integrated graphics is hard enough already, but it's even harder on 11 year old integrated graphics (and the low-end variant at that).

Calinou, are you familiar with the Sega Megadrive? the video processor, if I'm not mistaken, is about 12 megahertz. Here is an example of playing on it https://youtu.be/xhNv476tAs8
Intel HD Graphics 2000 with a frequency of 1000 megahertz, that is, almost a hundred times more powerful than Sega's video processor. The question is, where and what is 50% of the power of this GPU spent when displaying an empty screen of an empty Godot project? This is the power of as many as 50 such games as on Sega working in parallel. And is it possible to fix this by freeing up this power, and not wasting the power of the GPU in nowhere?

**Comment by mojoyup** [comment]
I JUST started getting about 90% GPU on my 1660TI with an R5 empty screen on 3.5.1 Steam Version. With my current scene I've been doing, it's easy 100% and my frames idle are ~40fps with nothing moving. It's usually 144fps. Just some extra confirmation. Started happening yesterday 1/15/23. Ironically at the SAME TIME, I just started getting transparent rendering priority issues, which have not existed since the beginning of the project 2 months ago. So I just finished setting ALL my transparent materials to render priorities > 0. NEVER had to do that before and I didn't change anything that is using them. So, something changed and I haven't fiddled with anything, just chugging along.

**Comment by RubberDuckShobe** [comment]
Unsure if I should open another issue for this, but I am having the same issue on Godot 4 instead of Godot 3. The issue suddenly manifested in a project I was working on, even in empty scenes. I tried creating a new project with an empty scene and the issue persisted.

This only happens when testing the game, the editor is fine. Capping the FPS doesn't help either.
My GPU is an NVIDIA RTX 2070.

**Comment by Calinou** [comment]
> Unsure if I should open another issue for this, but I am having the same issue on Godot 4 instead of Godot 3. The issue suddenly manifested in a project I was working on, even in empty scenes. I tried creating a new project with an empty scene and the issue persisted.
> 
> This only happens when testing the game, the editor is fine. Capping the FPS doesn't help either. My GPU is an NVIDIA RTX 2070.

How are you checking GPU utilization? The Windows task manager is wildly known to give unrealistic readouts. Try using something like NVIDIA's own overlay (<kbd>Alt + R</kbd>) instead.

**Comment by Calinou** [comment]
> Nevermind, it's definitely not just Task Manager, now the NVIDIA overlay shows extremely high usage as well:
> 
> ![grafik](https://user-images.githubusercontent.com/42943070/231501101-422b2067-d3f2-476d-9822-abceabec6685.png)

At uncapped FPS, this is expected. At capped FPS, high GPU utilization can occur if the GPU is downclocking as a result of low utilization (yes). Check the core and memory frequencies when running at capped FPS, as well as the power consumption figure (which is the figure you should care about the most).

**Comment by RubberDuckShobe** [comment]
Okay, so, mystery solved for real this time. Thanks for the additional info!
Turns out my issue was a combination of:
- Task Manager's questionable readouts, which makes the GPU usage appear much higher than it actually is
- Uncapped FPS, which made the GPU usage *actually* high

**Comment by Calinou** [comment]
@FyiurAmron Can you check if power consumption is actually high on your GPU with capped framerate/V-Sync, as mentioned in the above comments?

**Comment by FyiurAmron** [comment]
@Calinou Tested today with Godot_v3.5.2-stable_mono_win64

before running an empty project:
![image](https://user-images.githubusercontent.com/11016998/231615614-5dc085e6-0f75-4bde-ba3b-909b1b27cefa.png)
after runnning an empty project:
![image](https://user-images.githubusercontent.com/11016998/231615745-b7ec9939-318b-4b89-a988-b5f75c35aeb0.png)

The temperature slowly rises if it's left for longer periods of time. Afterwards, the fans go up. The problem is definitely organic and not an artifact of the measuring tools.

As I've already stated, neither manual framerate capping nor various VSync modes work here. I've tried using regular VSync via Godot, forced VSync via Radeon Software, I tried using Radeon Chill with various caps, and Radeon Enhanced Sync. I've measured 60 FPS all the way. I was also able to run with max FPS after disabling VSync, which increased the GPU temperatures and power consumption even more.

Like I said, the only things that worked were either switching to Godot 4, using `run/frame_delay_msec` or Low Processor Mode.

And BTW, as an aside: no, to reach 200W it actually requires more than just typical 100% GPU load. Even FurMark max GPU stress tests hit approx 150W avg power on my setup due to thermal throttling etc. I wouldn't call "200-210W" a "typical power consumption" for RX 580. Even 180W is the advertised max AFAIR, to go higher you need to both overclock it and select a thermal profile that would handle that much power. I'm able to run new-ish games at 1920x1080 at 30-60 FPS without much trouble or too much heating going on, staying <= 100W. I wouldn't say an empty Godot scene should about as demanding for my GPU than e.g. Titanfall 2, Against the Storm or Dyson Sphere Program at "High" quality settings :)

**Comment by FyiurAmron** [comment]
@Calinou I have an additional data point here. I went back to Godot dev after hiatus, so I went with 4.x which provides "Compatibility" renderer... lo and behold, as soon as I go with "Compatiblity", this bug is back! :D

Forward+ (everything in order)
![image](https://github.com/godotengine/godot/assets/11016998/2c13ab00-2ecb-423f-af65-27d9c4a0b717)

Mobile (everything in order)
![image](https://github.com/godotengine/godot/assets/11016998/be7e8bbf-11b1-48ac-bf0c-3467fdda517d)

Compatibility (yup, flat 100% GPU and _egg-cooking-on-the-PC-case mode enabled_ as soon as I switched to it)
![image](https://github.com/godotengine/godot/assets/11016998/d306d703-384f-4cb3-8dc2-3f214841479c)

But yeah, this is definitely linked to OpenGL code+drivers combo, and probably affects only a tiny minority of users.

**Comment by FyiurAmron** [comment]
> My problem even with VSync disabled has 100% GPU issues.

By _disabling_ VSync, you explicitly tell the GPU to use 100% of power and render as many frames as possible, ignoring the actual monitor refresh rate. To cap the GPU use, you usually _enable_ VSync or any of its relatives. Either you have a typo in your reply to me, or what you get is the actual expected result, no bug whatsoever.
